<a href="https://colab.research.google.com/github/Anny-Verly/Projeto-Python-MASL/blob/main/O_que_aprendi_Caderno_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#  <center>  </center>

# <center>  O que Aprendi: Caderno 3 </center> - Modelagem caixa branca e caixa preta de sistemas de primeira e de segunda ordem


<p align="justify">  O Caderno 2 apresenta uma discussão acerca de funções de transferências (FT's) e representações em Espaço de Estados (E.E). As principais funções utilizadas para gerar tais modelos, em Python, são apresentadas. Com o intuito de aplicar o conhecimento adquirido, o presente $notebook$ propõe os seguintes exercícios.


<a name="ex1"></a>
## **Exercício 1**

<p align="justify"> Um sistema de controle de posição converte um comando de entrada de posição em uma resposta de saída de posição. Os sistemas de controle de posição encontram uma vasta aplicação em antenas, braços robóticos e acionadores de disco de computadores \cite{nise}.

<p align="justify"> As figuras C2.1 e C2.2 apresentam, respectivamente, o esquema de funcionamento e o diagrama de blocos de um esquema simplificado de controle de posição. O objetivo é fazer com que a saída dada em graus angulares, $\theta_{s}$, siga o ângulo de entrada do potenciômetro, $\theta_{en}$. O potenciômetro de entrada converte o deslocamento angular, $\theta_{en}$, em uma tensão elétrica. Analogamente, o deslocamento angular da saída, $\theta_{s}$, é convertido em tensão elétrica por meio do potenciômetro da malha de realimentação. Amplificadores de sinal de potência ressaltam a diferença entre as tensões de entrada e de saída. Tal sinal de atuação amplificado aciona o conjunto motor/carga/engrenagens.

<center>   
<p align="center">
<img src="https://github.com/Anny-Verly/Projeto-Python-MASL/blob/main/Imagens_Caderno_2/Figura%20C2.1.jpg?raw=true"
width="800" height="" style="display: block; margin: 0 auto" />
</p>  Figura C2.1: Diagrama de blocos de um sistemas de controle de posição de ângulo de azimute. </center>

<center>   
<p align="center">
<img src="https://github.com/Anny-Verly/Projeto-Python-MASL/blob/main/Imagens_Caderno_2/Figura%20C2.2.png?raw=true"
width="800" height="" style="display: block; margin: 0 auto" />
</p>  Figura C2.2: Esquema de funcionamento de um sistema de controle de posição de ângulo de azimute. </center>


<p align="justify"> $\textbf{1.1)}$ Por meio de programação em Python, declare todas as funções de transferência (FTs) apresentadas no diagrama de blocos da Figura C2.2.


In [ ]:
pip install control

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 549.6/549.6 kB 6.8 MB/s eta 0:00:00


In [ ]:
#1.1) Respostas:
import math
import control as ctl
#1.1)Declarando as FT's:
s = ctl.tf('s')
#Potênciômetro de Entrada:
FT1= ctl.tf(1,math.pi)
print(FT1)
#Pré-Amplificador:
K=1
FT2= ctl.tf([K],1)
print(FT2)
#Aplificador de Potência:
FT3=ctl.tf([100],[1,100])
print(FT3)
#Motor, Carga e Engrenagens:
FT4=ctl.tf(0.2083,[1,1.71,0])
print(FT4)
#Potênciômetro de Saída:
FT5=ctl.tf(1,math.pi)
print(FT5)

<TransferFunction>: sys[0]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']


  1
-----
3.142

<TransferFunction>: sys[1]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']


1
-
1

<TransferFunction>: sys[2]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']


  100
-------
s + 100

<TransferFunction>: sys[3]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']


   0.2083
------------
s^2 + 1.71 s

<TransferFunction>: sys[4]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']


  1
-----
3.142




<p align="justify"> $\textbf{1.2)}$ Utilize as FT's declaradas no item 1.1 para realizar a redução do diagrama de blocos da Figura C2.2.

In [ ]:
FT6=FT2*FT3*FT4
print(FT6)
FT7=ctl.feedback(FT6,FT5,-1)
print(FT7)
FT_tot=FT7*FT1
print(FT_tot)

<TransferFunction>: sys[6]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']


         20.83
-----------------------
s^3 + 101.7 s^2 + 171 s

<TransferFunction>: sys[7]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']


                 65.44
---------------------------------------
3.142 s^3 + 319.5 s^2 + 537.2 s + 20.83

<TransferFunction>: sys[8]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']


               65.44
------------------------------------
9.87 s^3 + 1004 s^2 + 1688 s + 65.44



<a name="ex2"></a>
## **Exercício 2**

Utilizando programação em Python, converta a FT da planta obtida no [Exercício 1](#ex1) (item 1.2) em uma representação de espaço de estados. Depois, realize a conversão E.E --> FT.

In [ ]:
SS=ctl.tf2ss(FT_tot)
print(SS)
TF=ctl.ss2tf(SS)
print(TF)

<StateSpace>: sys[8]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']
States (3): ['x[0]', 'x[1]', 'x[2]']

A = [[-101.71       -171.           -6.63039493]
     [   1.            0.            0.        ]
     [   0.            1.            0.        ]]

B = [[1.]
     [0.]
     [0.]]

C = [[0.         0.         6.63039493]]

D = [[0.]]

<TransferFunction>: sys[10]
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']


     -1.421e-13 s + 6.63
------------------------------
s^3 + 101.7 s^2 + 171 s + 6.63



## **Exercício 3**

<p align="justify"> Obtenha um modelo de estado para o sistema físico não linear descrito pela equação (citar):


$$ y^{n}(t)=f_1[t,y(t),\dot{y}(t), ...,y^{n-1}(t),u(t) ] -f_2[t,y(t),\dot{y}(t), ...,y^{n-2}(t)]\dot{u}(t); $$

<p align="justify"> em que, $y(t)$ e $u(t)$ são, respectivamente saída e entrada (escalares).

<p align="justify"> $Sugestão$: Utilize como variável de estado a relação $ x_{n}=y^{n-1}-f_2[t,y,\dot{y}, ...,y^{n-2}]u.$

<p align="justify"> $\textbf{Resposta}$: Se $x_1(t)=y(t)$, $x_2(t)=\dot{y}(t)$, $x_3(t)=\ddot{y}(t)$, $x_4(t)=\dddot{y}(t)$, $\dots$ e $ x_{n}=y^{n-1}-f_2[t,y,\dot{y}, ...,y^{n-2}]u$,  então:

$\dot{x}_{1}(t)=x_{2}(t)$

$\dot{x}_{2}(t)=x_{3}(t)$

$\dot{x}_{3}(t)=x_{4}(t)$

$\vdots$

<p align="justify"> $\dot{x}_{n-1}(t)=x_{n}(t)= y^{n-1}-f_2[t,y,\dot{y}, ...,y^{n-2}]u$

<p align="justify"> $\dot{x}_{n}(t)= y^{n}-\dot{f}_2[t,y,\dot{y}, ...,y^{n-2}]u - {f}_2[t,y,\dot{y}, ...,y^{n-2}]\dot{u}$

ou ainda,


$\dot{x}_{n}(t)$ = $ f_1[t,y(t),\dot{y}(t), ...,y^{n-1}(t),u(t) ] -f_2[t,y(t),\dot{y}(t), ...,y^{n-2}(t)]\dot{u}(t) - \dot{f}_2[t,y,\dot{y}, ...,y^{n-2}]u - {f}_2[t,y,\dot{y}, ...,y^{n-2}]\dot{u}$

resultando em:


 $\dot{x}_{n}(t)$ = $ f_1[t,y(t),\dot{y}(t), ...,y^{n-1}(t),u(t) ] -2f_2[t,y(t),\dot{y}(t), ...,y^{n-2}(t)]\dot{u}(t) - \dot{f}_2[t,y,\dot{y}, ...,y^{n-2}]u$.





<a name="ex4"></a>
## **Exercício 4**

<p align="justify"> Considere o sistema representado pela equação diferencial, conhecida por equação de Van der Pol:


$$ m \ddot{x}+2𝜺(x^2-1)\dot{x} +kx=0, $$

<p align="justify"> em que onde ${\displaystyle x(t)}$ é a posição do oscilador e 𝜺  é um parâmetro que indica a não linearidade a força do  amortecimento. A equação de Van der Pol é utilizada para o estudo de osciladores com amortecimento não linear, por exemplo.

<p align="justify"> Considerando que o modelo de Van der Pol se trata de uma equação diferencial ordinária, autônoma, não linear de segunda ordem, pede-se:

<p align="justify"> 4.1) Desenvolva o modelo não linear em espaço de estados considerando: $x_1=x$ e $x_2=\dot{x}$.

<p align="justify"> $\textbf{Resposta}$: Se $x_1(t)=x(t)$, $x_2(t)=\dot{x}(t)$, então:

$\dot{x}_{1}(t)=x_2(t)$

$\dot{x}_{2}(t)= \frac{-kx_{1}(t) -2𝜺(x_1^2(t)-1)x_{2}(t)}{m} $

<p align="justify"> 4.2) Declare o modelo em E.E não linear, em Python, utilizando o comando $def$.

In [ ]:
def VanderPol(y,t,k,eps,m):
  x1,x2 =y
  dx1dt=x2
  dx2dt=(-k*(x1)-2*eps*(x1**2-1)*x2)/m
  return [dx1dt,dx2dt]

<a name="ex5"></a>
## **Exercício 5**

<p align="justify"> Considere a equação não linear que descreve o movimento do pêndulo simples:

$$ \ddot{\theta} +\frac{g}{\ell} sen(\theta)=0,$$

<p align="justify"> em que $\ell$ representa o comprimento  da haste do pêndulo e $g$ é a aceleração da gravidade. A partir da equação do movimento, declare em Python a representação em E.E equivalente.

<center>   
<p align="center">
<img src="https://github.com/Anny-Verly/Projeto-Python-MASL/blob/main/Imagens_Caderno_2/Figura%20C2.3.png?raw=true"
width="300" height="" style="display: block; margin: 0 auto" />
</p>  Figura C2.3: Representação de um pêndulo simples. </center>

<p align="justify"> $\textbf{Resposta}$: Se $x_1(t)=\theta(t)$ e $x_2(t)=\dot{\theta}(t)$, então:

$\dot{x}_{1}(t)=x_2(t)$

$\dot{x}_{2}(t)= -\frac{g}{\ell}sen(x_1(t)) $.

In [ ]:
import math
def VanderPol(y,t,g,ell):
  x1,x2 =y
  dx1dt=x2
  dx2dt=-g/ell*math.sin(x1)
  return [dx1dt,dx2dt]

**Exercício Proposto 3.1:**

Considere o circuito de um filtro passa faixa mostrado na Figura 19
Onde, $R_1 = 100Ω, R_2 = 1000Ω, C_1 = 3.10^{−6}F, C_2 = 1.10^{-6}F$.

Determine:

a) A Função de Transferência e o modelo em Espaço de Estados do circuito.

b) As respostas ao impulso e degrau utilizando a FT.

c) As respostas ao impulso e degrau utilizando o modelo em EE.

d) Compare os resultados das letras b e c.


**Exercício Proposto 3.3:**

Considere novamente a FT representada pela [Equação 3.16](#316). Realize a realimentação
unitária deste sistema e repita as letras a e b do [Exemplo 3.5](#ex35). Houve alguma diferença? Discuta os resultados obtidos.

**Exercício Proposto 3.4:**

Utilize as referências [[2]](#ref2), [[1]](#ref1) ou [[8]](#ref8) para estudar sobre a resposta de sistemas decontrole com polos adicionais.


**Exercício Proposto 3.2:**

Considere um sistema de controle representado pela [Equação 3.10](#eq310).

a) Com $𝜔_n$ = 2, plote a resposta ao degrau do sistema para cinco valores de ζ
compreendidos na região subamortecida. Utilize uma única tela e observe os resultados.

b) Com ζ = 0.8, plote a resposta ao degrau do sistema para $𝜔_n$ = 2, 4, 6, 8 e 10.

Utilize uma única tela e observe os resultados.

c) Com wn = 2, escolha um valor de ζ para cada um dos formatos de resposta
citados anteriormente. Utilize o subplot para plotar cada tipo de resposta em um gráfico diferente. Discuta os resultados.